osutil/disks/disks.go

```diff
@@ -0,0 +1,282 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2020 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package disks
+
+import (
+	"bufio"
+	"bytes"
+	"fmt"
+	"io"
+	"os/exec"
+	"path/filepath"
+	"strconv"
+	"strings"
+
+	"github.com/snapcore/snapd/osutil"
+)
+
+// diskFromMountPoint is exposed for mocking from other tests via
+// MockMountPointDisksToPartionMapping, but we can't just assign
+// diskFromMountPointImpl to diskFromMountPoint due to signature differences,
+// the former returns a *disk, the latter returns a Disk, and as such they can't
+// be assigned to each other
+var diskFromMountPoint = func(mountpoint string, opts *Options) (Disk, error) {
+	return diskFromMountPointImpl(mountpoint, opts)
+}
+
+// Options is a set of options used when querying information about
+// partition and disk devices.
+type Options struct {
+	// IsDecryptedDevice indicates that the mountpoint is referring to a
+	// decrypted device.
+	IsDecryptedDevice bool
+}
+
+// Disk is a single physical disk device that contains partitions.
+type Disk interface {
+	// FindMatchingPartitionUUID finds the partition uuid for a partition
+	// matching the specified label on the disk. Note that for non-ascii labels
+	// like "Some label", the label should be encoded using \x<hex> for
+	// potentially non-safe characters like in "Some\x20Label".
+	FindMatchingPartitionUUID(string) (string, error)
+
+	// MountPointIsFromDisk returns whether the specified mountpoint corresponds
+	// to a partition on the disk. Note that this only considers partitions
+	// and mountpoints found when the disk was identified with
+	// DiskFromMountPoint.
+	MountPointIsFromDisk(string, *Options) (bool, error)
+
+	// Dev returns the string "major:minor" number for the disk device.
+	Dev() string
+}
+
+func parseDeviceMajorMinor(s string) (int, int, error) {
+	errMsg := fmt.Errorf("invalid device number format: (expected <int>:<int>)")
+	devNums := strings.SplitN(s, ":", 2)
+	if len(devNums) != 2 {
+		return 0, 0, errMsg
+	}
+	maj, err := strconv.Atoi(devNums[0])
+	if err != nil {
+		return 0, 0, errMsg
+	}
+	min, err := strconv.Atoi(devNums[1])
+	if err != nil {
+		return 0, 0, errMsg
+	}
+	return maj, min, nil
+}
+
+var udevadmProperties = func(device string) ([]byte, error) {
+	cmd := exec.Command("udevadm", "info", "--query", "property", "--name", device)
```

Comment:
BTW. we have some code in interfaces hotplug that parses the output of `udevadm info --export-db`, which it the similar format to `udevadm info --name <device>`. Perhaps we could consider unifying the code in some later refactor?

Reply:
hmm perhaps it would make sense, but seems definitely TODO territory for now I think